### PR TITLE
Migrate Signatures to common-jvm from consent-signaling-client

### PR DESCRIPTION
**Pull request by jonmolle**
X509Certificate.verifySignature(signedData: SignedData) will
remain in consent-signaling-client as it is only used internally
in that repo and uses a repo-specific proto.

Tests migrated as well.

Followup PR to move the references.

Note that the extensions to StorageClient and
StorageClient.Blob are kept in a separate location to avoid
depending on the signing code.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)

### CODE REVIEW

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
I'm not so sure what I think about this migration. I'm not sure what makes consenting to an exchange workflow different than consenting to a measurement or is just the storage use case that doesn't fit in consent-signaling?

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-:-Mj6N5Nt0TLpyP4EvAlP:bul74at)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @corbantek, @efoxepstein, @jonmolle, @kgaynor, and @SanjayVas)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6MKj_9F7gbKsifTQG):*
i would think that `Flow<SignedData>` is a great candidate for storage - https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/9f3a39e9d0270a8e0f7db7a94740e581eb937f70/src/main/proto/wfa/measurement/api/v2alpha/exchange_step.proto#L63. 

What makes you think its repo-specific? Is was part of panel client api until recently.
___
*[src/main/kotlin/org/wfanet/measurement/common/crypto/Signatures.kt, line 81 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6Mw0F-BDtEy2Evpb5:-Mj6Mw0F-BDtEy2Evpb6:b8rsikh) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/e2b49795c392ac6dda416a5fdb5c1299564650a9/src/main/kotlin/org/wfanet/measurement/common/crypto/Signatures.kt#L81)):*
> ```Kotlin
>  */
> fun X509Certificate.verifySignedFlow(
>   data: Flow<ByteString>,
> ```

see my other comment about accepting a `Flow<SignedData>`

**Review by jonmolle — COMMENTED** [review]
Specifically the latter. It was requested that I move this out as (I was told that) the only reason Signatures wasn't already in common-jvm was that it was only used in consent-signaling-client. Now that it has a storage use case I was asked to move it out.

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-:-Mj6TZKg435QLC6L-83i:b-ddvor7)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @corbantek, @efoxepstein, @jonmolle, @kgaynor, @SanjayVas, and @stevenwarejones)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6MKj_9F7gbKsifTQG):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

i would think that `Flow<SignedData>` is a great candidate for storage - https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/9f3a39e9d0270a8e0f7db7a94740e581eb937f70/src/main/proto/wfa/measurement/api/v2alpha/exchange_step.proto#L63. 

What makes you think its repo-specific? Is was part of panel client api until recently.
</blockquote></details>

Sorry, you're absolutely right, my check for this was wrong. What I should have said (and what I found) was that common-jvm doesn't depend on cross-media-measurement-api and that (because I also chose to not use it) it seemed okay to leave X509Certificate.verifySignature(signedData: SignedData) in consent-signaling-client.

You know, I'm not entirely sure why that wouldn't work. I do plan to add a way to read chunked signed data (with each chunk and some metadata being signed as well) to better support distributed reads, but I was under the impression that just reading SignedData from a file had some issue (perhaps that you had to read it in entirely to find the sig?)

Working on this, let me get back to you with either a change or a better answer.
___
*[src/main/kotlin/org/wfanet/measurement/common/crypto/Signatures.kt, line 81 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6Mw0F-BDtEy2Evpb5:-Mj6W43k4-WtfE-6S0K6:b-cr19c) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/e2b49795c392ac6dda416a5fdb5c1299564650a9/src/main/kotlin/org/wfanet/measurement/common/crypto/Signatures.kt#L81)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

see my other comment about accepting a `Flow<SignedData>`
</blockquote></details>

Working on this, see your other comment.

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 5 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: 5 of 8 files reviewed, 3 unresolved discussions (waiting on @corbantek, @jonmolle, @kgaynor, @SanjayVas, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/storage/BUILD.bazel, line 18 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6gLTcCZpiphAF40NN:-Mj6gLTd1FOsqtEwwws6:b-7ilvfu) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/e2b49795c392ac6dda416a5fdb5c1299564650a9/src/main/kotlin/org/wfanet/measurement/storage/BUILD.bazel#L18)):*
> ```Starlark
> 
> kt_jvm_library(
>     name = "verified_storage_xtensions",
> ```

typo?

**Review by jonmolle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: 5 of 8 files reviewed, 3 unresolved discussions (waiting on @corbantek, @jonmolle, @kgaynor, @SanjayVas, and @stevenwarejones)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6MKj_9F7gbKsifTQG):*
<details><summary><i>Previously, jonmolle wrote…</i></summary><blockquote>

Sorry, you're absolutely right, my check for this was wrong. What I should have said (and what I found) was that common-jvm doesn't depend on cross-media-measurement-api and that (because I also chose to not use it) it seemed okay to leave X509Certificate.verifySignature(signedData: SignedData) in consent-signaling-client.

You know, I'm not entirely sure why that wouldn't work. I do plan to add a way to read chunked signed data (with each chunk and some metadata being signed as well) to better support distributed reads, but I was under the impression that just reading SignedData from a file had some issue (perhaps that you had to read it in entirely to find the sig?)

Working on this, let me get back to you with either a change or a better answer.
</blockquote></details>

Hi Steven, update. If we're passing a flow of SignedData objects there's no real need to support stream validation there (since each SignedData object is in memory). Again, I do plan to add support later for something like you're talking about specifically for encrypted queries. We can go over that with the team in a dev connect before you clear this if you prefer or I can add a TODO.

**Review by jonmolle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: 4 of 8 files reviewed, 3 unresolved discussions (waiting on @corbantek, @efoxepstein, @jonmolle, @kgaynor, @SanjayVas, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/storage/BUILD.bazel, line 18 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6gLTcCZpiphAF40NN:-MjAZAkX79iOiB-JUB2Q:bb6un1j) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/e2b49795c392ac6dda416a5fdb5c1299564650a9/src/main/kotlin/org/wfanet/measurement/storage/BUILD.bazel#L18)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

typo?
</blockquote></details>

Done

**Review by jonmolle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: 4 of 8 files reviewed, 3 unresolved discussions (waiting on @corbantek, @efoxepstein, @kgaynor, @SanjayVas, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/common/crypto/Signatures.kt, line 81 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6Mw0F-BDtEy2Evpb5:-MjAZERz1bAeamRl3iUq:bwpzbj4) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/e2b49795c392ac6dda416a5fdb5c1299564650a9/src/main/kotlin/org/wfanet/measurement/common/crypto/Signatures.kt#L81)):*
<details><summary><i>Previously, jonmolle wrote…</i></summary><blockquote>

Working on this, see your other comment.
</blockquote></details>

OK, consolidating discussion in the other comment.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: all files reviewed, 2 unresolved discussions (waiting on @corbantek, @kgaynor, @SanjayVas, and @stevenwarejones)

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 7 of 8 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: all files reviewed, 4 unresolved discussions (waiting on @corbantek, @jonmolle, @kgaynor, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/common/crypto/exception/BUILD.bazel, line 6 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-MjBXly_2s1UpGSJyGiD:-MjBXly_2s1UpGSJyGiE:b-2lwiwb) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/ddcb022159266ad2fe9b0310929e863554d2b12e/src/main/kotlin/org/wfanet/measurement/common/crypto/exception/BUILD.bazel#L6)):*
> ```Starlark
> 
> kt_jvm_library(
>     name = "exception",
> ```

Re: package structure, there are two types of package trees:

1. A package *encapsulates* its subpackages. In this case, a package may depend on its subpackages but other trees should only depend on the parent package.
2. Subpackages may *specialize* their parent packages. In this case, a parent package should never depend on any of its subpackages.
    * e.g. how the `storage` package defines the base `StorageClient` and the subpackages have specialized implementations of it.

This follows neither of those patterns. We tend to do (2), or even a strict subset of (2) where we only have code in leaf packages. This is why we have `common` packages.

Basically either the code that's currently in `crypto` should get pushed down into sibling subpackages of `exception`, or we pull exceptions up into the `crypto` package.
___
*[src/main/kotlin/org/wfanet/measurement/storage/VerifiedStorageExtensions.kt, line 15 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-MjBevSr9wGp1GIFqpXn:-MjBevSr9wGp1GIFqpXo:b-eqsw4a) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/ddcb022159266ad2fe9b0310929e863554d2b12e/src/main/kotlin/org/wfanet/measurement/storage/VerifiedStorageExtensions.kt#L15)):*
> ```
> // TODO: Move this file to common-jvm
> ```
Drop obsolete TODO

**Review by jonmolle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: 1 of 8 files reviewed, 4 unresolved discussions (waiting on @corbantek, @efoxepstein, @kgaynor, @SanjayVas, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/storage/VerifiedStorageExtensions.kt, line 15 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-MjBevSr9wGp1GIFqpXn:-MjBy-0q6RICwR1TCMF9:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/ddcb022159266ad2fe9b0310929e863554d2b12e/src/main/kotlin/org/wfanet/measurement/storage/VerifiedStorageExtensions.kt#L15)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> ```
> // TODO: Move this file to common-jvm
> ```
Drop obsolete TODO
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/measurement/common/crypto/exception/BUILD.bazel, line 6 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-MjBXly_2s1UpGSJyGiD:-MjBy0RvCZR3tjp8df_p:btfygke) ([raw file](https://github.com/world-federation-of-advertisers/common-jvm/blob/ddcb022159266ad2fe9b0310929e863554d2b12e/src/main/kotlin/org/wfanet/measurement/common/crypto/exception/BUILD.bazel#L6)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Re: package structure, there are two types of package trees:

1. A package *encapsulates* its subpackages. In this case, a package may depend on its subpackages but other trees should only depend on the parent package.
2. Subpackages may *specialize* their parent packages. In this case, a parent package should never depend on any of its subpackages.
    * e.g. how the `storage` package defines the base `StorageClient` and the subpackages have specialized implementations of it.

This follows neither of those patterns. We tend to do (2), or even a strict subset of (2) where we only have code in leaf packages. This is why we have `common` packages.

Basically either the code that's currently in `crypto` should get pushed down into sibling subpackages of `exception`, or we pull exceptions up into the `crypto` package.
</blockquote></details>

I took the second of your two options to minimize changes. Done.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: all files reviewed, 2 unresolved discussions (waiting on @corbantek, @kgaynor, and @stevenwarejones)

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 1 of 8 files at r1, 6 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @corbantek and @kgaynor)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6MKj_9F7gbKsifTQG):*
<details><summary><i>Previously, jonmolle wrote…</i></summary><blockquote>

Hi Steven, update. If we're passing a flow of SignedData objects there's no real need to support stream validation there (since each SignedData object is in memory). Again, I do plan to add support later for something like you're talking about specifically for encrypted queries. We can go over that with the team in a dev connect before you clear this if you prefer or I can add a TODO.
</blockquote></details>

Good point.

**Review by corbantek — APPROVED** [review]
:lgtm:

> Reviewed 1 of 8 files at r1, 6 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-:-MjG3Eud6zHYiVsHjlmi:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kgaynor)

**Review by jonmolle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kgaynor)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/common-jvm/34#-Mj6MKj_9F7gbKsifTQG):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

Good point.
</blockquote></details>

Done (infeasible for now, just triggering reviewable). So as a second update, the goal is apparently to have common- repos not depend on anything in wfa. If we want to validate SignedData protos outside of consent-signaling in the short term I think the solution is just to invoke verifySignature(data, sig) with the fields of SignedData. It isn't like the verifySignature that takes SignedData even is unit tested. It seems like moving proto definitions around to maintain the dependency tree is a bigger pain that the benefit here (at least for now).
